src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -109,51 +117,132 @@ public void WriteResponseHeaders(int streamId, int statusCode, IHeaderDictionary
                         _outgoingFrame.ContinuationFlags = Http2ContinuationFrameFlags.END_HEADERS;
                     }
 
-                    Append(_outgoingFrame.Raw);
+                    _outputWriter.Write(_outgoingFrame.Raw);
                 }
             }
         }
 
-        public Task WriteDataAsync(int streamId, ReadOnlySpan<byte> data, CancellationToken cancellationToken)
-            => WriteDataAsync(streamId, data, endStream: false, cancellationToken: cancellationToken);
-
-        public Task WriteDataAsync(int streamId, ReadOnlySpan<byte> data, bool endStream, CancellationToken cancellationToken)
+        public Task WriteDataAsync(int streamId, Http2StreamFlowControl flowControl, ReadOnlySequence<byte> data, bool endStream)
         {
-            var tasks = new List<Task>();
+            // The Length property of a ReadOnlySequence can be expensive, so we cache the value.
+            long dataLength = data.Length;
 
             lock (_writeLock)
             {
-                _outgoingFrame.PrepareData(streamId);
-
-                while (data.Length > _outgoingFrame.Length)
+                if (flowControl.IsAborted)
                 {
-                    data.Slice(0, _outgoingFrame.Length).CopyTo(_outgoingFrame.Payload);
-                    data = data.Slice(_outgoingFrame.Length);
+                    return Task.CompletedTask;
+                }
 
-                    tasks.Add(WriteAsync(_outgoingFrame.Raw, cancellationToken));
+                // Zero-length data frames are allowed to be sent immediately even if there is no space available in the flow control window.
+                // https://httpwg.org/specs/rfc7540.html#rfc.section.6.9.1
+                if (dataLength != 0 && dataLength > flowControl.Available)
+                {
+                    return WriteDataAsyncAwaited(streamId, flowControl, data, dataLength, endStream);
                 }
 
-                _outgoingFrame.Length = data.Length;
+                flowControl.Advance(dataLength, out _);
```

Comment:
Does this warrant an overload that verifies these assumptions internally? E.g. that does not a return an awaitable and Asserts Available?